### PR TITLE
Initial MMU Implementation & MMIO Death

### LINKED ISSUE #5: [feat] Add support for virtual memory
We want memory pages, we want virtual memory.

Tutorial: 
- https://github.com/bztsrc/raspi3-tutorial/tree/master/10_virtualmemory
- https://jsandler18.github.io/tutorial/wrangling-mem.html

POSIX compatibility:
- `mmap()`: a bit more complex than expected. Should also be able to map to memory some files.
- `unmap()`

Linux compatibility:
- `brk()` and `sbrk()` system call (see https://linux.die.net/man/2/sbrk)

### LINKED ISSUE #35: [feat] Add function to resolve aliases in DeviceTree
We need to access aliases multiple times when parsing DeviceTree. Actually, we need first to query `/aliases/{something}` and then again with the previous result.

### LINKED ISSUE #36: [feat] Add functions to parse common properties types in Property of DeviceTree

### LINKED ISSUE #37: Use DeviceTree in all modules instead of MMIO

### LINKED ISSUE #38: Add support for the different UARTs
Add an interface that allow to choose the UART to use. The hardware support many UARTs (4-5), it will be cool to be able to use multiple of them for debugging.

### LINKED ISSUE #39: [bug] Fix clang
Current problems:
- `char buffer[21] = {0}` in debug.cpp -> PROBLEM, we lost the execution thread
- Error when constructing PropertyMessage in device.cpp Device::init()

### PR DESCRIPTION
**Pull request by desfreng**
This will be my PR for :
- Implementation of an initial Memory Mapping
- Death of the MMIO, replaced by Device Tree functions
- Kernel Testing in Release mode

Test are done on:
- A Raspberry Pi 3b+
- A Compute Module 4 ! (for now it works :partying_face:  !)

### CODE REVIEW

File: kernel/mmu.c [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+  // finally, toggle some bits in system control register to enable page translation
+  {
+    uint64_t r;
+    asm volatile("dsb ish; isb; mrs %0, sctlr_el1" : "=r"(r));
+    r |= 0xC00800;      // set mandatory reserved bits
+    r &= ~((1 << 25) |  // clear EE, little endian translation tables
+           (1 << 24) |  // clear E0E
+           (1 << 19) |  // clear WXN
+           (1 << 12) |  // clear I, no instruction cache
```
Comment (hgruniaux) [inline kernel/mmu.c]: Why disabling the instruction cache?

File: kernel/mmu.c [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+           (1 << 4) |   // clear SA0
+           (1 << 3) |   // clear SA
+           (1 << 2) |   // clear C, no cache at all
```
Comment (hgruniaux) [inline kernel/mmu.c]: Same here, why disabling the cache?

**Comment by hgruniaux** [comment]
Tested on Raspberry Pi 4, it works.

Also, I have tested my branch on release (optimizations enabled) and it doesn't works.

**Comment by desfreng** [comment]
Tested on CM4 and RPi 3b+.
Both Release and Debug with GCC 13.2.0 and Clang 17.0.6.

**Comment by desfreng** [comment]
I'll do the Process Memory API in another PR, lot's of changes mist be merge here first I think.

**Review by hgruniaux — APPROVED** [review]

**Pull request merged**